larpandoracontent/LArHelpers/LArHitWidthHelper.cc

```diff
@@ -0,0 +1,411 @@
+  
+LArHitWidthHelper::ClusterToParametersMapStore* LArHitWidthHelper::ClusterToParametersMapStore::m_instance{nullptr};
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+LArHitWidthHelper::ClusterToParametersMapStore* LArHitWidthHelper::ClusterToParametersMapStore::Instance()
+{
+    if(!m_instance)
+        m_instance = new ClusterToParametersMapStore();
+    return m_instance;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+LArHitWidthHelper::ClusterToParametersMap* LArHitWidthHelper::ClusterToParametersMapStore::GetMap()
+{
+    return &m_clusterToParametersMap;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+LArHitWidthHelper::ConstituentHitVector LArHitWidthHelper::GetConstituentHits(const Cluster *const pCluster, const float maxConstituentHitWidth, const float hitWidthScalingFactor) 
+{
+  
+    OrderedCaloHitList orderedCaloHitList(pCluster->GetOrderedCaloHitList());
+
+    if (orderedCaloHitList.empty())
+        throw StatusCodeException(STATUS_CODE_NOT_FOUND);
+
+    ConstituentHitVector constituentHitVector;
+
+    for (OrderedCaloHitList::const_iterator iter = orderedCaloHitList.begin(); iter != orderedCaloHitList.end(); ++iter)
+    {
+        for(CaloHitList::const_iterator hitIter = iter->second->begin(); hitIter != iter->second->end(); ++hitIter) 
+        {
+	  
+            const CaloHit *const pCaloHit = *hitIter;
+
+            const float hitWidth = pCaloHit->GetCellSize1() * hitWidthScalingFactor;
+            const unsigned int numberOfConstituentHits = floor(hitWidth/maxConstituentHitWidth) + 1;
+            const float constituentHitWidth = hitWidth/static_cast<float>(numberOfConstituentHits);
+	    
+	        // start at end of cluster with the lowest x value
+            float xPositionAlongHit(pCaloHit->GetPositionVector().GetX() - (hitWidth/2));
+	        for(unsigned int i(0); i < numberOfConstituentHits; ++i) 
+             {
+                 i == 0 ? xPositionAlongHit += constituentHitWidth/2 : xPositionAlongHit += constituentHitWidth;
+                 CartesianVector consituentHitPosition(xPositionAlongHit, 0, pCaloHit->GetPositionVector().GetZ());
+
+                constituentHitVector.push_back(ConstituentHit(consituentHitPosition, constituentHitWidth, pCluster));
+             }
+        }
+    }
+    return constituentHitVector;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+LArHitWidthHelper::ConstituentHitVector LArHitWidthHelper::GetUniformConstituentHits(const Cluster *const pCluster, const float constituentHitWidth, const float hitWidthScalingFactor)
```

Comment:
GetConstituentHits splits a hit into constituent hits such that no hit length is added to the hit. The constituent hits have a weight between 0 and maxConstituentHitWidth. The constituent hits within the same hit have the SAME constituent hit width but the constituent hit width between hits will be DIFFERENT. This is fine when it comes to fitting because a weighted fit is performed.

GetUniformHits splits a hit into uniform constituent hits with constituent hit widths equal to hitWidth and so the overall hit width is padded. This ensures that each constituent hit has the same width (and therefore fitting weight) even between hits. I had to do this when I used the sliding linear fit.